### ISSUE TITLE
make Error 2 --- error: ‘class web::websockets::client::websocket_client_config’ has no member named ‘set_ssl_context_callback’

### ISSUE THREAD

**Issue by CRYPTO-7CA** [sony/nmos-cpp#121]
I tried to build the Sony NMOS CPP project on my end and I have followed the provided instructions as much as I understood. At some point before we managed to build it successfully but lately we are struggling getting it built. My platform is Linux (Centos 7). 

During the MAKE stage of this project I got stuck and can't figure out what went wrong. At this moment I have cloned this repository locally with the latest commit ID being https://github.com/sony/nmos-cpp/commit/bf6e13c28729089468f90dff3958d243eb820f49

Here's what I see when trying to build it:

```
Scanning dependencies of target nmos-cpp_static
[ 74%] Building CXX object CMakeFiles/nmos-cpp_static.dir/cpprest/api_router.cpp.o
[ 74%] Building CXX object CMakeFiles/nmos-cpp_static.dir/cpprest/host_utils.cpp.o
[ 75%] Building CXX object CMakeFiles/nmos-cpp_static.dir/cpprest/http_utils.cpp.o
[ 75%] Building CXX object CMakeFiles/nmos-cpp_static.dir/cpprest/json_escape.cpp.o
[ 75%] Building CXX object CMakeFiles/nmos-cpp_static.dir/cpprest/json_storage.cpp.o
[ 76%] Building CXX object CMakeFiles/nmos-cpp_static.dir/cpprest/json_utils.cpp.o
[ 76%] Building CXX object CMakeFiles/nmos-cpp_static.dir/cpprest/json_validator_impl.cpp.o
[ 76%] Building CXX object CMakeFiles/nmos-cpp_static.dir/cpprest/json_visit.cpp.o
[ 77%] Building CXX object CMakeFiles/nmos-cpp_static.dir/cpprest/ws_listener_impl.cpp.o
[ 77%] Building CXX object CMakeFiles/nmos-cpp_static.dir/nmos/admin_ui.cpp.o
[ 77%] Building CXX object CMakeFiles/nmos-cpp_static.dir/nmos/api_downgrade.cpp.o
[ 78%] Building CXX object CMakeFiles/nmos-cpp_static.dir/nmos/api_utils.cpp.o
[ 78%] Building CXX object CMakeFiles/nmos-cpp_static.dir/nmos/channels.cpp.o
[ 78%] Building CXX object CMakeFiles/nmos-cpp_static.dir/nmos/client_utils.cpp.o
/home/voronezh/Downloads/nmos-cpp/Development/nmos/client_utils.cpp: In function ‘web::websockets::client::websocket_client_config nmos::make_websocket_client_config(const settings&)’:
/home/voronezh/Downloads/nmos-cpp/Development/nmos/client_utils.cpp:75:16: error: ‘class web::websockets::client::websocket_client_config’ has no member named ‘set_ssl_context_callback’
         config.set_ssl_context_callback(details::make_client_ssl_context_callback<web::websockets::client::websocket_exception>(settings));
                ^
In file included from /home/voronezh/Downloads/nmos-cpp/Development/nmos/client_utils.cpp:7:0:
/home/voronezh/Downloads/nmos-cpp/Development/cpprest/details/system_error.h: In instantiation of ‘ExceptionType web::details::from_boost_system_system_error(const boost::system::system_error&) [with ExceptionType = web::http::http_exception]’:
/home/voronezh/Downloads/nmos-cpp/Development/nmos/client_utils.cpp:44:88:   required from ‘nmos::details::make_client_ssl_context_callback(const settings&) [with ExceptionType = web::http::http_exception; nmos::settings = web::json::value]::__lambda144’
/home/voronezh/Downloads/nmos-cpp/Development/nmos/client_utils.cpp:34:21:   required from ‘struct nmos::details::make_client_ssl_context_callback(const settings&) [with ExceptionType = web::http::http_exception; nmos::settings = web::json::value]::__lambda144’
/home/voronezh/Downloads/nmos-cpp/Development/nmos/client_utils.cpp:46:13:   required from ‘std::function<void(boost::asio::ssl::context&)> nmos::details::make_client_ssl_context_callback(const settings&) [with ExceptionType = web::http::http_exception; nmos::settings = web::json::value]’
/home/voronezh/Downloads/nmos-cpp/Development/nmos/client_utils.cpp:60:118:   required from here
/home/voronezh/Downloads/nmos-cpp/Development/cpprest/details/system_error.h:71:71: error: no matching function for call to ‘web::http::http_exception::http_exception(std::error_code, const char*)’
             return ExceptionType(to_std_error_code(e.code()), e.what());
                                                                       ^
/home/voronezh/Downloads/nmos-cpp/Development/cpprest/details/system_error.h:71:71: note: candidates are:
In file included from /usr/local/include/cpprest/http_client.h:50:0,
                 from /home/voronezh/Downloads/nmos-cpp/Development/nmos/client_utils.h:4,
                 from /home/voronezh/Downloads/nmos-cpp/Development/nmos/client_utils.cpp:1:
/usr/local/include/cpprest/http_msg.h:246:5: note: web::http::http_exception::http_exception(int, const std::error_category&)
     http_exception(int errorCode, const std::error_category& cat) : m_errorCode(std::error_code(errorCode, cat))
     ^
/usr/local/include/cpprest/http_msg.h:246:5: note:   no known conversion for argument 1 from ‘std::error_code’ to ‘int’
/usr/local/include/cpprest/http_msg.h:222:5: note: web::http::http_exception::http_exception(int, const string_t&)
     http_exception(int errorCode, const utility::string_t& whatArg)
     ^
/usr/local/include/cpprest/http_msg.h:222:5: note:   no known conversion for argument 1 from ‘std::error_code’ to ‘int’
/usr/local/include/cpprest/http_msg.h:212:5: note: web::http::http_exception::http_exception(int)
     http_exception(int errorCode) : m_errorCode(utility::details::create_error_code(errorCode))
     ^
/usr/local/include/cpprest/http_msg.h:212:5: note:   candidate expects 1 argument, 2 provided
/usr/local/include/cpprest/http_msg.h:197:5: note: web::http::http_exception::http_exception(const string_t&)
     http_exception(const utility::string_t& whatArg) : m_msg(utility::conversions::to_utf8string(whatArg)) {}
     ^
/usr/local/include/cpprest/http_msg.h:197:5: note:   candidate expects 1 argument, 2 provided
/usr/local/include/cpprest/http_msg.h:190:7: note: web::http::http_exception::http_exception(const web::http::http_exception&)
 class http_exception : public std::exception
       ^
/usr/local/include/cpprest/http_msg.h:190:7: note:   candidate expects 1 argument, 2 provided
/usr/local/include/cpprest/http_msg.h:190:7: note: web::http::http_exception::http_exception(web::http::http_exception&&)
/usr/local/include/cpprest/http_msg.h:190:7: note:   candidate expects 1 argument, 2 provided
In file included from /home/voronezh/Downloads/nmos-cpp/Development/nmos/client_utils.cpp:7:0:
/home/voronezh/Downloads/nmos-cpp/Development/cpprest/details/system_error.h: In function ‘ExceptionType web::details::from_boost_system_system_error(const boost::system::system_error&) [with ExceptionType = web::http::http_exception]’:
/home/voronezh/Downloads/nmos-cpp/Development/cpprest/details/system_error.h:72:9: warning: control reaches end of non-void function [-Wreturn-type]
         }
         ^
make[2]: *** [CMakeFiles/nmos-cpp_static.dir/nmos/client_utils.cpp.o] Error 1
make[1]: *** [CMakeFiles/nmos-cpp_static.dir/all] Error 2
make: *** [all] Error 2
```

From what I understood something is not right with finding the websocket dependencies, that we built on Microsoft's CPP Rest SDK.

Here's the command I performed just before "make"-ing it:

```
cmake3 .. \
  -DCMAKE_BUILD_TYPE:STRING="Release" \
  -DBOOST_INCLUDEDIR:PATH="/home/voronezh/Downloads/boost_1_67_0" \
  -DBOOST_LIBRARYDIR:PATH="/home/voronezh/Downloads/boost_1_67_0/lib" \
  -DWEBSOCKETPP_INCLUDE_DIR:PATH="/home/voronezh/Downloads/cpprestsdk/Release/libs/websocketpp"
```

I got the following dependencies installed: 

- boost_1_67_0
- cmake-3.12.3
- cpprestsdk (commit ID=8f0393dd7be08f2d9b44cc09b8b97f9bafb109a5)
- mDNSResponder-878.200.35
- openssl-OpenSSL_1_1_1c

Did I do something incorrect or is the build broken?

**Comment by garethsb-sony** [comment]
> I got the following dependencies installed:
>
> * boost_1_67_0
> * cmake-3.12.3
> * cpprestsdk (commit ID=8f0393dd7be08f2d9b44cc09b8b97f9bafb109a5)
> * mDNSResponder-878.200.35
> * openssl-OpenSSL_1_1_1c
>
> Did I do something incorrect or is the build broken?

The cpprestsdk commit 8f0393dd7be08f2d9b44cc09b8b97f9bafb109a5 is version 2.10.9.

The minimum required version is 2.10.11 (which introduced the change that is preventing your compilation). The most well-tested version is 2.10.14.

Hope that helps!

**Comment by CRYPTO-7CA** [comment]
That solved this problem. We had outdated instructions on our end.
Then I met my next problem with OpenSSL having a conflict.

```
[ 97%] Building CXX object CMakeFiles/nmos-cpp-test.dir/sdp/test/sdp_test.cpp.o
[ 98%] Linking CXX executable nmos-cpp-test
/usr/bin/ld: warning: libcrypto.so.10, needed by /usr/lib64/libssl.so, may conflict with libcrypto.so.1.1
/usr/bin/ld: CMakeFiles/nmos-cpp-test.dir/nmos/test/api_utils_test.cpp.o: undefined reference to symbol 'OPENSSL_add_all_algorithms_noconf@@libcrypto.so.10'
//usr/lib64/libcrypto.so.10: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [nmos-cpp-test] Error 1
make[1]: *** [CMakeFiles/nmos-cpp-test.dir/all] Error 2
make: *** [all] Error 2
```

Which could be solved by providing this parameter as part of the cmake3 command (I got it from a guardian angel). I have yet to test it.

`-DOPENSSL_ROOT_DIR="$HOME/Downloads/openssl-OpenSSL_1_1_1c" `

I would encourage you to update the instructions in particular with the updated OpenSSL in mind please when "Cmaking" the NMOS CPP project. I don't see it in the dependencies list nor do I see build instructions for it (although they are available online). It would be good to include some build instructions for OpenSSL as well. It will reduce such obstacles. I can provide these for Linux (tested on CentOS 7).

```
sudo yum install libtool perl-core zlib-devel -y
cd $HOME/Downloads
curl -O -L https://github.com/openssl/openssl/archive/OpenSSL_1_1_1c.tar.gz
tar -zxf OpenSSL_1_1_1c.tar.gz
cd openssl-OpenSSL_1_1_1c
./config --prefix=/usr/local/openssl --openssldir=/usr/local/openssl shared zlib
make
make test
sudo make install
```

Thanks!

**Comment by garethsb-sony** [comment]
The [dependency on OpenSSL](https://github.com/sony/nmos-cpp/blob/master/Documents/Dependencies.md#openssl) is clearly mentioned in the docs.

I think you're correct that on some platforms the naming of OpenSSL binaries has changed, so CMake and build instructions could  be usefully added.

**Comment by CRYPTO-7CA** [comment]
I should have been more clear. I meant in this **Dependencies** summary list OpenSSL is not being mentioned
![image](https://user-images.githubusercontent.com/61016421/74532299-eca52000-4f3f-11ea-8d98-3a0ad58ca8b8.png)

Building the NMOS CPP did work now. Here's the working CMAKE command used before "making" it.
```
cmake .. \
  -DCMAKE_BUILD_TYPE:STRING="Release" \
  -DWEBSOCKETPP_INCLUDE_DIR:PATH="$HOME/Downloads/cpprestsdk/Release/libs/websocketpp" \
  -DBOOST_INCLUDEDIR:PATH="$HOME/Downloads/boost_1_67_0" \
  -DOPENSSL_ROOT_DIR="$HOME/Downloads/openssl-OpenSSL_1_1_1c" 
```
I encourage you to update the instructions for CMaking the NMOS CPP.

For OpenSSL the OS CentOS 8 comes shipped with OpenSSL 1.1.1c by default but for building CPP Rest SDK the package `openssl-devel` was required on my end.

I consider my case being solved.
Thanks!

**Comment by garethsb-sony** [comment]
>  I meant in this **Dependencies** summary list OpenSSL is not being mentioned

It's mentioned in the screenshot list too, under C++ REST SDK. 🙄

Thanks for the hints on the CMake incantation that was required on your platform.

**Comment by CRYPTO-7CA** [comment]
Ah, I misinterpreted that one. It's a sub-dependency. Still may require additional installation steps depending on which OS is being used.
All good now. I'm closing this one.

**Comment by gururajsb13** [comment]
Hi @garethsb,

We are using the Conan package manager for managing dependencies instead of installing them separately. My understanding is that Conan will pull the correct versions of dependencies during the build process. I noticed that you have committed an updated conanfile.txt in the Development branch.

However, after cloning the latest source from the master branch and attempting to build it, I am still encountering the same compilation error as mentioned earlier. Is there anything specific I need to do before starting the build?

Thank you for your help.

Best regards,
Guru

**Comment by garethsb** [comment]
Hi @gururajsb13,

Probably better to open a new issue, and include full details of the environment you are building in, your build steps, and the error message you are seeing.

**Comment by lo-simon** [comment]
Hi @gururajsb13, please ensure you have Conan 2 installed and all the old Conan caches removed before retrying. see https://github.com/sony/nmos-cpp/blob/master/Documents/Installation-with-Conan.md

**Comment by gururajsb13** [comment]
Thanks for the quick reply. 

@lo-simon, I updated the Conan version to 2.0.17, but that didn't help.
@garethsb, sure I'll open new issue and include more details.
